src/guards/slice_memory_guard.rs

```diff
@@ -50,6 +50,60 @@ impl<'a, T> SliceMemoryGuard<'a, T> {
         }
         SliceMemoryGuard { memory }
     }
+
+    /// Initialize memory guard using given iterator.
+    /// Automatically shrink's memory to given items' count.
+    /// `Ok(guard)` will be returned in this case.
+    ///
+    /// If items' count is too large to place in memory, moves it into new `Vec` and continue collecting into it.
+    /// `Err(vec)` will be returned in this case.
+    #[inline]
+    pub unsafe fn new_from_iter(memory: &'a mut [MaybeUninit<T>], source: impl Iterator<Item = T>) -> Result<Self, Vec<T>> {
+        let size_hint = source.size_hint();
+        let mut iter = source.peekable();
+
+        // Fulfilling placed memory
+        for (index, item) in memory.into_iter().enumerate() {
+            match iter.next() {
+                // While iterator returns new value, write it
+                Some(value) => {
+                    write(item.as_mut_ptr(), value);
+                },
+                // When it returns None then slicing memory and returning the guard
+                None => {
+                    return Ok(SliceMemoryGuard {
+                        memory: &mut memory[0..index],
+                    });
+                },
+            }
+        }
+
+        if iter.peek().is_some() {
```

Comment:
Do we really need a peekable here? Without it, we could just grab that element, and push it to the vector after we copy the initial stuff.